### PR TITLE
8332387: [lworld] monitorenter on value class instances must throw IdentityException instead of IllegalMonitorStateException

### PR DESCRIPTION
**Pull request by fparain**
Change the exception thrown by monitorenter when attempted on an instance of a value class.
Monitorexit is not changed according to some discussions about the JVMS changes (it still throws IllegalMonitorStateException).

Tested with Mach5, tier1-3.

Fred

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8332387](https://bugs.openjdk.org/browse/JDK-8332387): [lworld] monitorenter on value class instances must throw IdentityException instead of IllegalMonitorStateException (**Bug** - P3)

### Reviewers
 * [Dan Heidinga](https://openjdk.org/census#heidinga) (@DanHeidinga - no project role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/1108/head:pull/1108` \
`$ git checkout pull/1108`

Update a local copy of the PR: \
`$ git checkout pull/1108` \
`$ git pull https://git.openjdk.org/valhalla.git pull/1108/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1108`

View PR using the GUI difftool: \
`$ git pr show -t 1108`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/1108.diff">https://git.openjdk.org/valhalla/pull/1108.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/1108#issuecomment-2125652368)

### CODE REVIEW

File: src/hotspot/share/c1/c1_LIR.cpp [hunk 1]
```diff
@@ -838,7 +838,7 @@ void LIR_OpVisitState::visit(LIR_Op* op) {
       assert(opLock->_result->is_illegal(), "unused");
 
       do_stub(opLock->_stub);
-      do_stub(opLock->_throw_imse_stub);
+      do_stub(opLock->_throw_ie_stub);
```
Comment (DanHeidinga) [inline src/hotspot/share/c1/c1_LIR.cpp:841]: Does `_throw_ie_stub` apply to both `lir_lock` & 'lir_unlock`?  I thought only `monitorenter` was being changed.

**Review by DanHeidinga — COMMENTED** [review]
Looks straightforward and correct by pattern match =) apart from a minor question

File: src/hotspot/share/c1/c1_LIR.cpp (hunk 1 above)
Comment (fparain) [reply src/hotspot/share/c1/c1_LIR.cpp:841]: When creating a LIR_OpLock for lir_unlock (in LIR_List::unlock_object(), c1_LIR.cpp:1576), no stub is provided so _throw_ie_stub is null. And do_stub() does nothing if the pointer passed in argument is null.
Did I miss something?

Comment (DanHeidinga) [reply src/hotspot/share/c1/c1_LIR.cpp:841]: You're correct. Thanks for confirming

**Review by DanHeidinga — APPROVED** [review]

**Comment by fparain** [comment]
Dan,

Thank you for the review.

**Pull request closed without merging**